cypress-io/cypress
# feat: introduce v8 snapshots to improve startup performance

- Closes #18480

### User facing changelog


Incorporated V8 snapshots into the build process of the electron binary. This improves startup time and reduces the time to download and unzip the binary during installation.

### Additional details


In order to improve start up time, Cypress uses [electron mksnapshot](https://github.com/electron/mksnapshot) for generating [V8 snapshots](https://v8.dev/blog/custom-startup-snapshots) for both development and prod. These snapshots are generated as a post install step locally and as an additional step in CI. The generation process:

1. Traverses the server code
2. Modifies it to be snapshottable
3. Creates a bundled file that contains logic to hydrate either the exports or definitions of all server dependencies
4. Runs `electron-mksnapshot` to generate the snapshot file from the bundled file in step 3
5. Copies the snapshot file from step 4 to the Cypress electron distribution where it is used by electron on start up to hydrate the dependencies relatively instantaneously

This feature adds 5 new packages:

* `@packages/v8-snapshot-require` - Tool to load a snapshot for Electron applications that was created by `@tooling/v8-snapshot`.
* `@packages/packherd-require` - Loads modules that have been bundled by `@tooling/packherd`
* `@tooling/electron-mksnapshot` - A rewrite of [electron/mksnapshot](https://github.com/electron/mksnapshot) to support multiple versions
* `@tooling/packherd` - Herds all dependencies reachable from an entry and packs them.
* `@tooling/v8-snapshot` - Tool to create a snapshot for Electron applications.

### Steps to test


Install the module/binary using the instructions on the last commit on the PR, run the app, and ensure that nothing is broken. There's no new functionality so everything should be working as it did previously.

### How has the user experience changed?


Startup time has improved. Here's an example of the type of improvement seen in cypress-realworld-app:

| Architecture | Open Mode - Baseline | Open Mode - With V8 Snapshots |
| ------------ | -------------------- | ----------------------------- |
| Darwin       | 1.6s                 | 0.5s                          |
| Linux        | 2.2s                 | 0.6s                          |
| Windows      | 3.7s                 | 0.6s                          |

### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
